lib/usd/hdMaya/delegates/proxyDelegate.cpp

```diff
@@ -0,0 +1,376 @@
+
+void SetupPluginCallbacks() {
+    MStatus status;
+
+    isALPluginLoaded.store(IsALPluginLoaded());
+
+    // Set up callback to notify of plugin load
+    TF_DEBUG(HDMAYA_AL_CALLBACKS)
+        .Msg("HdMayaProxyDelegate - creating PluginLoaded callback\n");
+    MSceneMessage::addStringArrayCallback(
+        MSceneMessage::kAfterPluginLoad, PluginCallback, nullptr, &status);
+    TF_VERIFY(status, "Could not set pluginLoaded callback");
+
+    // Set up callback to notify of plugin unload
+    TF_DEBUG(HDMAYA_AL_CALLBACKS)
+        .Msg("HdMayaProxyDelegate - creating PluginUnloaded callback\n");
+    MSceneMessage::addStringArrayCallback(
+        MSceneMessage::kAfterPluginUnload, PluginCallback, nullptr, &status);
+    TF_VERIFY(status, "Could not set pluginUnloaded callback");
+}
+
+} // namespace
+
+HdMayaProxyDelegate::HdMayaProxyDelegate(const InitData& initData)
+    : HdMayaDelegate(initData) {
+    TF_DEBUG(HDMAYA_AL_PROXY_DELEGATE)
+        .Msg(
+            "HdMayaProxyDelegate - creating with delegateID %s\n",
+            GetMayaDelegateID().GetText());
+
+    MStatus status;
+
+#if WANT_UFE_BUILD
+    if (usdUfeRtid == 0) {
+        try {
+            usdUfeRtid =
+                UFE_NS::RunTimeMgr::instance().getId(USD_UFE_RUNTIME_NAME);
+        }
+        // This shoudl catch ufe's InvalidRunTimeName exception, but they don't
+        // expose that!
+        catch (...) {
+            TF_WARN("USD UFE Runtime plugin not loaded!\n");
+        }
+    }
+#endif // WANT_UFE_BUILD
+}
+
+HdMayaProxyDelegate::~HdMayaProxyDelegate() {
+    TF_DEBUG(HDMAYA_AL_PROXY_DELEGATE)
+        .Msg(
+            "HdMayaProxyDelegate - destroying with delegateID %s\n",
+            GetMayaDelegateID().GetText());
+}
+
+HdMayaDelegatePtr HdMayaProxyDelegate::Creator(const InitData& initData) {
+    static std::once_flag setupPluginCallbacksOnce;
+    std::call_once(setupPluginCallbacksOnce, SetupPluginCallbacks);
+
+    if (!isALPluginLoaded.load()) { return nullptr; }
+    return std::static_pointer_cast<HdMayaDelegate>(
+        std::make_shared<HdMayaProxyDelegate>(initData));
+}
+
+void HdMayaProxyDelegate::AddAdapter(HdMayaProxyAdapter* adapter) {
+    std::lock_guard<std::mutex> lock(_allAdaptersMutex);
+    _allAdapters.insert(adapter);
+}
+
+void HdMayaProxyDelegate::RemoveAdapter(HdMayaProxyAdapter* adapter) {
+    std::lock_guard<std::mutex> lock(_allAdaptersMutex);
+    _allAdapters.erase(adapter);
+}
+
+void HdMayaProxyDelegate::Populate() {
+    // Does nothing - delegate exists only for PreFrame and
+    // PopulateSelectedPaths
+}
+
+void HdMayaProxyDelegate::PreFrame(const MHWRender::MDrawContext& context) {
+    std::lock_guard<std::mutex> lock(_allAdaptersMutex);
+    for (auto adapter : _allAdapters) { adapter->PreFrame(); }
+}
+
+void HdMayaProxyDelegate::PopulateSelectedPaths(
+    const MSelectionList& mayaSelection, SdfPathVector& selectedSdfPaths,
+    const HdSelectionSharedPtr& selection) {
+#ifdef KXL_TO_FINISH
```

Comment:
Here we have code which isn't finished. We added support for UFE selection, but in order to get selection working without UFE we need a new interface on MayaUsdProxyShapeBase to return proxy->selectedPaths() + proxy->selectionList() (one should do, unless I'm missing something).

FWIW. We don't want to explore MayaUsdProxyShapeBase with too many interfaces, so first would be good to know if this is important for you?

Reply:
So, we actually have another branch, which we haven't pushed yet, that also dealt with selection in non-ufe code.  Formerly, it required changes in both the AL repo and the hdmaya repo - now, of course, they can be merged now that they're both under one roof.  However, it looks like the current state of things is that the AL-code is still in an unpushed branch, but the hdmaya code was still in here - protected under a `HD_MAYA_AL_OVERRIDE_PROXY_SELECTION` #define… I've stripped this code for now:

https://github.com/Autodesk/maya-usd/pull/231/commits/8acb4e77809f714266430fd41142e8fee950fcf3

As for the current state of affairs - though I'd like to eventually get everyone using UFE in production, it might be a bit before that happens... and I think selection is important enough that we should provide something in the meantime.

So - the question is, do we want to get non-UFE selection working as a requirement for this PR, or make it a separate issue / PR for after this?  My main focus of this PR was to migrate all the code / features currently in refactoring_sandbox - so I guess I would vote to add it later.  But I'm fine with either decision.

Reply:
Indeed, the purpose is to migrate the code from refactoring_sandbox. Non-UFE selection is out of scope for this PR.

Reply:
Ok - so leave as is (with the KXL_TO_FINISH) for now, then?

Reply:
Unless you know it's the next thing you will do after this PR - My preference would be to remove it since it currently dead code. Otherwise, remove to something more descriptive.

Reply:
Ok, removed it - just left a small TODO in the header